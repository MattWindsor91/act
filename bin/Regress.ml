(* This file is part of 'act'.

   Copyright (c) 2018 by Matt Windsor

   Permission is hereby granted, free of charge, to any person
   obtaining a copy of this software and associated documentation
   files (the "Software"), to deal in the Software without
   restriction, including without limitation the rights to use, copy,
   modify, merge, publish, distribute, sublicense, and/or sell copies
   of the Software, and to permit persons to whom the Software is
   furnished to do so, subject to the following conditions:

   The above copyright notice and this permission notice shall be
   included in all copies or substantial portions of the Software.

   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
   EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
   MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
   NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
   BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
   ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
   CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
   SOFTWARE. *)

open Core
open Lib
open Utils

type spec =
  { cvars : string list
  }
[@@deriving sexp]
;;

let find_spec specs (path : Fpath.t) =
  let file = Fpath.to_string path in
  file
  |> List.Assoc.find specs ~equal:(String.Caseless.equal)
  |> Result.of_option
    ~error:(
      Error.create_s
        [%message "File mentioned in spec is missing" ~file]
    )
;;

let regress_run_asm ((module L) : (module Asm_job.Runner))
    (path : Fpath.t) mode specs passes (file : Fpath.t) =
  let open Or_error.Let_syntax in

  let%bind spec = find_spec specs file in

  Fmt.pr "## %a\n\n```\n@." Fpath.pp file;

  let input =
    { Asm_job.inp = Io.In_source.file Fpath.(path // file)
    ; outp = Io.Out_sink.stdout
    ; passes
    ; symbols = spec.cvars
    }
  in

  let%map _ = match mode with
    | `Litmusify -> L.litmusify input
    | `Explain   -> L.explain input
  in

  Out_channel.flush stdout;
  printf "```\n";
;;

let read_specs path =
  let spec_file = Fpath.(path / "spec") in
  Or_error.try_with
    (fun () ->
       Sexp.load_sexp_conv_exn (Fpath.to_string spec_file)
         [%of_sexp: (string, spec) List.Assoc.t]
    )
;;

let diff_to_error = function
  | First file ->
    Or_error.error_s
      [%message "File mentioned in test spec, but doesn't exist"
          ~file
      ]
  | Second file ->
    Or_error.error_s
      [%message "File in test directory, but doesn't have a spec"
          ~file
      ]
;;

let check_files_against_specs specs (test_paths : Fpath.t list) =
  let spec_set = specs |> List.map ~f:fst |> String.Set.of_list in
  let files_set =
    test_paths
    |> List.map ~f:Fpath.to_string
    |> String.Set.of_list
  in
  String.Set.symmetric_diff spec_set files_set
  |> Sequence.map ~f:diff_to_error
  |> Sequence.to_list
  |> Or_error.combine_errors_unit
;;

let regress_run_asm_many modename mode passes test_path =
  let open Or_error.Let_syntax in

  printf "# %s tests\n\n" modename;

  let emits = ["x86"; "att"] in
  let path = Fpath.(v test_path / "asm" / "x86" / "att") in
  let%bind l = Language_support.asm_runner_from_emits emits in
  let%bind specs = read_specs path in
  let%bind test_files = Io.Dir.get_files ~ext:"s" path in
  let%bind () = check_files_against_specs specs test_files in
  let results = List.map test_files
      ~f:(regress_run_asm l path mode specs passes)
  in
  let%map () = Or_error.combine_errors_unit results in
  printf "\nRan %d test(s).\n\n" (List.length test_files)
;;

let regress_explain =
  regress_run_asm_many "Explainer" `Explain
    (Sanitiser_pass.explain)
;;

let regress_litmusify =
  regress_run_asm_many "Litmusifier" `Litmusify
    (Sanitiser_pass.standard)
;;

let regress test_path =
  let open Or_error.Let_syntax in
  let%bind () = regress_explain test_path in
  let%map  () = regress_litmusify test_path in
  Out_channel.newline stdout
;;

let command =
  let open Command.Let_syntax in
  Command.basic
    ~summary:"runs regression tests"
    [%map_open
      let test_path = anon ("TEST_PATH" %: string) in
       fun () ->
         let o = Output.make ~verbose:false ~warnings:true in
         regress test_path |> Output.print_error o
    ]
;;
