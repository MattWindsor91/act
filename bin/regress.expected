# Explainer tests

## clang_macho_o3_xchg.s

```

_P0:  # <-- label (unused label, program boundary)
pushl %EBP # <-- stack (manipulates stack)
movl %ESP, %EBP # <-- move (manipulates stack)
movl $2, %EAX # <-- move
xchgl %EAX, _x # <-- RMW
movl $1, %EAX # <-- move
xchgl %EAX, _y # <-- RMW
popl %EBP # <-- stack (manipulates stack)
retl # <-- return
_P1:  # <-- label (unused label, program boundary)
pushl %EBP # <-- stack (manipulates stack)
movl %ESP, %EBP # <-- move (manipulates stack)
movl $2, %EAX # <-- move
xchgl %EAX, _y # <-- RMW
movl $1, %EAX # <-- move
xchgl %EAX, _x # <-- RMW
popl %EBP # <-- stack (manipulates stack)
retl # <-- return
```
## gcc_elf_o0_stackchk.s

```

P0:  # <-- label (unused label, program boundary)
pushl %EBP # <-- stack (manipulates stack)
movl %ESP, %EBP # <-- move (manipulates stack)
subl $24, %ESP # <-- arith (manipulates stack)
movl %GS:20, %EAX # <-- move
movl %EAX, -12(%EBP) # <-- move
xorl %EAX, %EAX # <-- logical
movl $2, -20(%EBP) # <-- move
movl -20(%EBP), %EAX # <-- move
movl %EAX, y # <-- move
mfence # <-- fence
movl x, %EAX # <-- move
movl %EAX, -16(%EBP) # <-- move
movl -16(%EBP), %EAX # <-- move
movl %EAX, t0r0 # <-- move
movl -12(%EBP), %EAX # <-- move
xorl %GS:20, %EAX # <-- logical
je .L2 # <-- jump
call __stack_chk_fail # <-- call
.L2:  # <-- label
leave # <-- call
ret # <-- return
P1:  # <-- label (unused label, program boundary)
pushl %EBP # <-- stack (manipulates stack)
movl %ESP, %EBP # <-- move (manipulates stack)
subl $24, %ESP # <-- arith (manipulates stack)
movl %GS:20, %EAX # <-- move
movl %EAX, -12(%EBP) # <-- move
xorl %EAX, %EAX # <-- logical
movl $1, -20(%EBP) # <-- move
movl -20(%EBP), %EAX # <-- move
movl %EAX, x # <-- move
mfence # <-- fence
movl $1, -16(%EBP) # <-- move
movl -16(%EBP), %EAX # <-- move
movl %EAX, y # <-- move
mfence # <-- fence
movl -12(%EBP), %EAX # <-- move
xorl %GS:20, %EAX # <-- logical
je .L4 # <-- jump
call __stack_chk_fail # <-- call
.L4:  # <-- label
leave # <-- call
ret # <-- return
```
## gcc_macho_o0_deref_cmpxchg.s

```

_P0:  # <-- label (unused label, program boundary)
pushl %EBP # <-- stack (manipulates stack)
movl %ESP, %EBP # <-- move (manipulates stack)
subl $16, %ESP # <-- arith (manipulates stack)
movl $0, _t0r0 # <-- move
movl $_x, -4(%EBP) # <-- move
movl $2, -8(%EBP) # <-- move
movl -8(%EBP), %EAX # <-- move
movl %EAX, %EDX # <-- move
movl -4(%EBP), %ECX # <-- move
movl _t0r0, %EAX # <-- move
lock cmpxchgl %EDX, (%ECX) # <-- RMW
movl %EAX, %EDX # <-- move
sete %AL # <-- ??
testb %AL, %AL # <-- ??
jne L3 # <-- jump
movl %EDX, _t0r0 # <-- move
L3:  # <-- label
leave # <-- call
ret # <-- return
_P1:  # <-- label (unused label, program boundary)
pushl %EBP # <-- stack (manipulates stack)
movl %ESP, %EBP # <-- move (manipulates stack)
subl $16, %ESP # <-- arith (manipulates stack)
movl $_x, -4(%EBP) # <-- move
movl $1, -8(%EBP) # <-- move
movl -8(%EBP), %EAX # <-- move
movl %EAX, %EDX # <-- move
movl -4(%EBP), %EAX # <-- move
movl %EDX, (%EAX) # <-- move
leave # <-- call
ret # <-- return
```
## gcc_macho_o0_deref_write.s

```

_P0:  # <-- label (unused label, program boundary)
pushl %EBP # <-- stack (manipulates stack)
movl %ESP, %EBP # <-- move (manipulates stack)
subl $16, %ESP # <-- arith (manipulates stack)
movl $_y, -4(%EBP) # <-- move
movl -4(%EBP), %EAX # <-- move
movl (%EAX), %EAX # <-- move
movl %EAX, -8(%EBP) # <-- move
movl -8(%EBP), %EAX # <-- move
movl %EAX, _t0r0 # <-- move
movl _t0r0, %EAX # <-- move
cmpl $1, %EAX # <-- compare
jne L3 # <-- jump
movl $1, _x # <-- move
L3:  # <-- label
leave # <-- call
ret # <-- return
_P1:  # <-- label (unused label, program boundary)
pushl %EBP # <-- stack (manipulates stack)
movl %ESP, %EBP # <-- move (manipulates stack)
subl $16, %ESP # <-- arith (manipulates stack)
movl _x, %EAX # <-- move
movl %EAX, _t1r0 # <-- move
movl _t1r0, %EAX # <-- move
cmpl $1, %EAX # <-- compare
jne L6 # <-- jump
movl $_y, -4(%EBP) # <-- move
movl $1, -8(%EBP) # <-- move
movl -8(%EBP), %EAX # <-- move
movl %EAX, %EDX # <-- move
movl -4(%EBP), %EAX # <-- move
movl %EDX, (%EAX) # <-- move
L6:  # <-- label
leave # <-- call
ret # <-- return
```
## gcc_macho_o0_immediate_heap.s

```

_P0:  # <-- label (unused label, program boundary)
pushl %EBP # <-- stack (manipulates stack)
movl %ESP, %EBP # <-- move (manipulates stack)
subl $16, %ESP # <-- arith (manipulates stack)
movl $_x, -4(%EBP) # <-- move
movl -4(%EBP), %EAX # <-- move
movl (%EAX), %EAX # <-- move
movl %EAX, -12(%EBP) # <-- move
movl -12(%EBP), %EAX # <-- move
movl %EAX, _t0r0 # <-- move
movl $_y, -8(%EBP) # <-- move
movl $1, -16(%EBP) # <-- move
movl -16(%EBP), %EAX # <-- move
movl %EAX, %EDX # <-- move
movl -8(%EBP), %EAX # <-- move
movl %EDX, (%EAX) # <-- move
leave # <-- call
ret # <-- return
_P1:  # <-- label (unused label, program boundary)
pushl %EBP # <-- stack (manipulates stack)
movl %ESP, %EBP # <-- move (manipulates stack)
subl $16, %ESP # <-- arith (manipulates stack)
movl $_y, -4(%EBP) # <-- move
movl $2, -12(%EBP) # <-- move
movl -12(%EBP), %EAX # <-- move
movl %EAX, %EDX # <-- move
movl -4(%EBP), %EAX # <-- move
movl %EDX, (%EAX) # <-- move
movl $_x, -8(%EBP) # <-- move
movl $1, -16(%EBP) # <-- move
movl -16(%EBP), %EAX # <-- move
movl %EAX, %EDX # <-- move
movl -8(%EBP), %EAX # <-- move
movl %EDX, (%EAX) # <-- move
leave # <-- call
ret # <-- return
```
## gcc_macho_o0_moves.s

```

_P0:  # <-- label (unused label, program boundary)
pushl %EBP # <-- stack (manipulates stack)
movl %ESP, %EBP # <-- move (manipulates stack)
subl $16, %ESP # <-- arith (manipulates stack)
movl _y, %EAX # <-- move
movl %EAX, -4(%EBP) # <-- move
movl -4(%EBP), %EAX # <-- move
movl %EAX, _t0r0 # <-- move
movl _t0r0, %EAX # <-- move
cmpl $1, %EAX # <-- compare
jne L3 # <-- jump
movl $1, _x # <-- move
L3:  # <-- label
leave # <-- call
ret # <-- return
_P1:  # <-- label (unused label, program boundary)
pushl %EBP # <-- stack (manipulates stack)
movl %ESP, %EBP # <-- move (manipulates stack)
subl $16, %ESP # <-- arith (manipulates stack)
movl _x, %EAX # <-- move
movl %EAX, _t1r0 # <-- move
movl _t1r0, %EAX # <-- move
cmpl $1, %EAX # <-- compare
jne L6 # <-- jump
movl $1, -4(%EBP) # <-- move
movl -4(%EBP), %EAX # <-- move
movl %EAX, _y # <-- move
mfence # <-- fence
L6:  # <-- label
leave # <-- call
ret # <-- return
```

Ran 6 test(s).

# Litmusifier tests

## clang_macho_o3_xchg.s

```

X86 clang_macho_o3_xchg.s

{ x = 0; y = 0; }

P0            | P1            ;
movl EAX, 2   | movl EAX, 2   ;
xchg [x], EAX | xchg [y], EAX ;
movl EAX, 1   | movl EAX, 1   ;
xchg [y], EAX | xchg [x], EAX ;

locations [x; y]
```
## gcc_elf_o0_stackchk.s

```

X86 gcc_elf_o0_stackchk.s

{
    t0gGSd200 = 0;
    t0gGSd201 = 0;
    t0r0 = 0;
    t0sZM12 = 0;
    t0sZM16 = 0;
    t0sZM20 = 0;
    t1gGSd200 = 0;
    t1gGSd201 = 0;
    t1sZM12 = 0;
    t1sZM16 = 0;
    t1sZM20 = 0;
    x = 0;
    y = 0;
}

P0                    | P1                    ;
movl EAX, [t0gGSd200] | movl EAX, [t1gGSd200] ;
movl [t0sZM12], EAX   | movl [t1sZM12], EAX   ;
xor EAX, EAX          | xor EAX, EAX          ;
movl [t0sZM20], 2     | movl [t1sZM20], 1     ;
movl EAX, [t0sZM20]   | movl EAX, [t1sZM20]   ;
movl [y], EAX         | movl [x], EAX         ;
mfence                | mfence                ;
movl EAX, [x]         | movl [t1sZM16], 1     ;
movl [t0sZM16], EAX   | movl EAX, [t1sZM16]   ;
movl EAX, [t0sZM16]   | movl [y], EAX         ;
movl [t0r0], EAX      | mfence                ;
movl EAX, [t0sZM12]   | movl EAX, [t1sZM12]   ;
xor EAX, [t0gGSd201]  | xor EAX, [t1gGSd201]  ;

locations [t0gGSd200; t0gGSd201; t0r0; t0sZM12; t0sZM16; t0sZM20; t1gGSd200;
           t1gGSd201; t1sZM12; t1sZM16; t1sZM20; x; y]
```
## gcc_macho_o0_deref_cmpxchg.s

```

X86 gcc_macho_o0_deref_cmpxchg.s

{ ZUt0r0 = 0; t0sZM4 = 0; t0sZM8 = 0; x = 0; }

P0                      | P1         ;
movl [ZUt0r0], 0        | mov [x], 1 ;
movl [t0sZM4], x        |            ;
movl [t0sZM8], 2        |            ;
movl EAX, [t0sZM8]      |            ;
movl EDX, EAX           |            ;
movl ECX, [t0sZM4]      |            ;
movl EAX, [ZUt0r0]      |            ;
lock cmpxchg [ECX], EDX |            ;
movl EDX, EAX           |            ;
sete AL                 |            ;
testb AL, AL            |            ;
jne L3                  |            ;
movl [ZUt0r0], EDX      |            ;
L3:                     |            ;

locations [ZUt0r0; t0sZM4; t0sZM8; x]
```
## gcc_macho_o0_deref_write.s

```

X86 gcc_macho_o0_deref_write.s

{ ZUt0r0 = 0; ZUt1r0 = 0; t0sZM8 = 0; x = 0; y = 0; }

P0                 | P1                 ;
mov EAX, [y]       | movl EAX, [x]      ;
movl [t0sZM8], EAX | movl [ZUt1r0], EAX ;
movl EAX, [t0sZM8] | movl EAX, [ZUt1r0] ;
movl [ZUt0r0], EAX | cmp EAX, 1         ;
movl EAX, [ZUt0r0] | jne L6             ;
cmp EAX, 1         | mov [y], 1         ;
jne L3             | L6:                ;
movl [x], 1        |                    ;
L3:                |                    ;

locations [ZUt0r0; ZUt1r0; t0sZM8; x; y]
```
## gcc_macho_o0_immediate_heap.s

```

X86 gcc_macho_o0_immediate_heap.s

{ ZUt0r0 = 0; t0sZM12 = 0; x = 0; y = 0; }

P0                  | P1         ;
mov EAX, [x]        | mov [y], 2 ;
movl [t0sZM12], EAX | mov [x], 1 ;
movl EAX, [t0sZM12] |            ;
movl [ZUt0r0], EAX  |            ;
mov [y], 1          |            ;

locations [ZUt0r0; t0sZM12; x; y]
```
## gcc_macho_o0_moves.s

```

X86 gcc_macho_o0_moves.s

{ ZUt0r0 = 0; ZUt1r0 = 0; t0sZM4 = 0; t1sZM4 = 0; x = 0; y = 0; }

P0                 | P1                 ;
movl EAX, [y]      | movl EAX, [x]      ;
movl [t0sZM4], EAX | movl [ZUt1r0], EAX ;
movl EAX, [t0sZM4] | movl EAX, [ZUt1r0] ;
movl [ZUt0r0], EAX | cmp EAX, 1         ;
movl EAX, [ZUt0r0] | jne L6             ;
cmp EAX, 1         | movl [t1sZM4], 1   ;
jne L3             | movl EAX, [t1sZM4] ;
movl [x], 1        | movl [y], EAX      ;
L3:                | mfence             ;
                   | L6:                ;

locations [ZUt0r0; ZUt1r0; t0sZM4; t1sZM4; x; y]
```

Ran 6 test(s).


